feat(codebuild): attribute-based compute type for Fleet

badmintoncryer opened the pull request:
### Issue # (if applicable)

None

### Reason for this change

AWS Codebuild supports for creating Fleet with attribute based compute type.
https://docs.aws.amazon.com/codebuild/latest/userguide/fleets.html#fleets.attribute-compute

You can specify minimum vCPU, disk and memory sizes. Codebuild automatically selects the instance type based on the compute configuration.

### Description of changes

Add `computeConfiguraion` prop to `FleetProps`.

### Description of how you validated changes

Add both unit and integ tests.

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation reviewed (changes requested):
The pull request linter has failed. See the aws-cdk-automation comment below for failure reasons. If you believe this pull request should receive an exemption, please comment and provide a justification.


A comment requesting an exemption should contain the text `Exemption Request`. Additionally, if clarification is needed add `Clarification Request` to a comment.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to a README file.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 605fd1639692f034dfedd62ccaaa6d9943491156
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0304f50a-b1ba-4ae8-8434-cf8dc966d610%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 8ee1d1c2433f9520c3cfc66609135385dcf4852b
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0cff7101-4801-410d-8d48-1c0e859f547b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 4ed820d92979aff52f7d34442edf6f2fc62c3132
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2a1232c5-5ef6-46f2-b33d-42a79105828c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 47d1de276666c3d39c7728bf0c9e9f3a29875866
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=342c2d58-067f-406b-a579-209eb9a01161%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: c8f21efc8d5828e9b2de8cdc4b3c9fb1345eff06
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ff3178c0-99a8-45a8-b368-315bd66a7351%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: b895761879ef2f5dee440651e35cca724f20d78b
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8f118e52-36d9-42ce-b303-65041aad7515%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: decf3a3075d69827b0ce1109170971e7be888e5f
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2291c49f-8814-4b02-b203-54b31b9edecc%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 52b4ef6f7a06e86424ef31e6f021a18f74b06268
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=94427c72-2126-454a-b995-962e6578cb3c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 13591193fd3435770db3777528b03c8b3fb862bd
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9d1fd2a4-d2b6-4bcd-8043-d0ef03c8cb8c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 8d1beada17f09f37a65f96b68c5142df79904bd6
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0550e8ba-7f0a-4521-8631-47cd7b71b632%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: d978959ac8ebb4d76e71dda86b50fe0d36fd3395
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=31e210fd-f389-4420-a0e1-743a20117133%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 382571da125279c1d48cff2a353111164f241137
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=369f2289-3337-498b-bf47-597695036ba4%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: fffcf4775fa075bd5e19cdc6d25f887d9381a669
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d82d1068-1639-4b19-9602-5cbb7564f390%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nmussy commented:
I swear I searched for issues and PRs before starting #32574 🤦 I'll close mine and give yours a review today

nmussy commented on packages/aws-cdk-lib/aws-codebuild/lib/compute-type.ts:
```diff
@@ -14,4 +14,5 @@ export enum ComputeType {
   LAMBDA_4GB = 'BUILD_LAMBDA_4GB',
   LAMBDA_8GB = 'BUILD_LAMBDA_8GB',
   LAMBDA_10GB = 'BUILD_LAMBDA_10GB',
+  ATTRIBUTE_BASED_COMPUTE = 'ATTRIBUTE_BASED_COMPUTE',
```
`FleetComputeType.ATTRIBUTE_BASED_COMPUTE` feels a little redundant, I went with:

```suggestion
  ATTRIBUTE_BASED = ComputeType.ATTRIBUTE_BASED_COMPUTE,
```

nmussy commented on packages/aws-cdk-lib/aws-codebuild/lib/fleet.ts:
```diff
@@ -138,13 +196,37 @@ export class Fleet extends Resource implements IFleet {
       computeType: props.computeType,
       environmentType: props.environmentType,
+      computeConfiguration: props.computeConfiguration ? {
+        disk: props.computeConfiguration.disk?.toGibibytes(),
+        machineType: props.computeConfiguration.machineType,
+        memory: props.computeConfiguration.memory?.toGibibytes(),
+        vCpu: props.computeConfiguration.vCpu,
```
Despite what the CloudFormation documentation says, the numeric props are not optional, see https://github.com/nmussy/aws-cdk/blob/6120d8f54103f81c4ec601221c1f56e5a4549236/packages/%40aws-cdk-testing/framework-integ/test/aws-codebuild/test/integ.project-fleet-attribute-based-compute.ts#L67-L73

Could you give them a default value, like [this for example](https://github.com/aws/aws-cdk/pull/32574/files/6120d8f54103f81c4ec601221c1f56e5a4549236#diff-5425be2f47cbad47acd51ad8eaa07758e9765df8bb28f242cade9f85f6c3d665R222-R224), and add an integ with just `{ machineType: GENERAL }`?

nmussy commented on packages/aws-cdk-lib/aws-codebuild/lib/fleet.ts:
```diff
@@ -138,13 +196,37 @@ export class Fleet extends Resource implements IFleet {
       throw new Error('baseCapacity must be greater than or equal to 1');
     }
 
+    if (props.computeConfiguration) {
```
I also had this validation to make sure the user saw how `ATTRIBUTE_BASED_COMPUTE` worked, leaving it up to you if you want to add it as well:

```suggestion
    if (props.computeType === FleetComputeType.ATTRIBUTE_BASED &&
      (!props.computeConfiguration || Object.keys(props.computeConfiguration).length === 0)) {
      throw new Error('At least one compute configuration criteria must be specified if computeType is "ATTRIBUTE_BASED"');
    }
    
    if (props.computeConfiguration) {
```

nmussy commented on packages/aws-cdk-lib/aws-codebuild/lib/fleet.ts:
You're missing a getter for the `fromFleetArn` `Import` class:

```ts
public get computeConfiguration(): FleetComputeConfiguration | undefined {
  throw new Error('Cannot retrieve computeConfiguration property from an imported Fleet');
}
```

nmussy reviewed (changes requested):
Had a quick first look by comparing our implementations, I'll try again with a more neutral outlook later. Thanks!

nmussy commented on packages/aws-cdk-lib/aws-codebuild/README.md:
```diff
@@ -490,6 +490,27 @@ new codebuild.Project(this, 'Project', {
+
+You can use [attribute-based compute](https://docs.aws.amazon.com/codebuild/latest/userguide/fleets.html#fleets.attribute-compute) for your fleet by setting the `computeType` to `ATTRIBUTE_BASED_COMPUTE`.
+This allows you to specify the attributes in `computeConfiguration` such as vCPUs, memory, disk space, and the machineType.
+After specifying one or some of the available attributes, CodeBuild will choose a compute type from the available supported instance types as the finalized `computeConfiguration`.
```
Something like this, sounds a little clunky but adds more information imo
```suggestion
After specifying some or all of the available attributes, CodeBuild will select the cheapest compute type from available instance types as that at least matches all given criteria.
```

nmussy commented on packages/aws-cdk-lib/aws-codebuild/README.md:
````diff
@@ -490,6 +490,27 @@ new codebuild.Project(this, 'Project', {
 })
 ```
 
+### Attribute based compute
````
The hyphen is also missing from the PR title, I think other instances are correct
```suggestion
### Attribute-based compute
```

nmussy reviewed (changes requested):
Additional nits

nmussy commented:
Something I was lacking as well was the ability to communicate to the user the exact instance type that will be selected. From what I could tell, it's only given out by the web console, and cannot be retrieved via the API, prior or post fleet creation. Even if we reverse engineered the selection algorithm, we would encounter the following issues:

* The availability of instance type is region dependent, see [docs](https://docs.aws.amazon.com/codebuild/latest/userguide/build-env-ref-compute-types.html#environment-reserved-capacity.types)
* I'm pretty certain the instance types are also account dependent and require contacting the AWS support, as I was getting [these kinds of error](https://github.com/nmussy/aws-cdk/blob/6120d8f54103f81c4ec601221c1f56e5a4549236/packages/%40aws-cdk-testing/framework-integ/test/aws-codebuild/test/integ.project-fleet-attribute-based-compute.ts#L54-L65)
* The web console selection algorithm is incorrect, or at least differs from what the CloudFormation instance selection is, see [this example](https://github.com/nmussy/aws-cdk/blob/6120d8f54103f81c4ec601221c1f56e5a4549236/packages/%40aws-cdk-testing/framework-integ/test/aws-codebuild/test/integ.project-fleet-attribute-based-compute.ts#L42-L52)

So overall, not great DX, but not much we can do from our end

badmintoncryer commented:
Thanks @nmussy! I'll update my PR later as I'm a bit busy right now.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 14842ccf19da3802fa87c99d3a9e8195a6c3941d
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=205f64bc-3b39-44fa-a7c0-7f39feb411c6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 43b9bfadcf6d2ae674a7df7ae434e4a7813cf229
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8522aff7-3ba1-429e-ad00-1c5d9422e475%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 06080d51188d83c328a34582f8c01683566c53ee
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=509fc159-d95a-49d6-9a2c-1a479f486a20%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

badmintoncryer commented:
@nmussy Thank you very much for your thorough review and detailed feedback. I believe I have addressed all the points you mentioned. I plan to check later whether the CI passes successfully.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: cc122af3605c76cd00cd44792f92d17af50f9df9
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=22cb03bb-ee26-4094-9943-e59f07fb4338%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 483bedbf89643410d519226c790a99d344021035
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e6f29e0c-2e48-4a1f-9a1f-c72503653c36%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 83da08e216203cc20504bfbf6d5229b9f940eaa4
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c9ade6d3-1374-4279-a5b5-d7ccb0923a7a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 30f86e7252ce8514cab118db5e6d2aeb344e1a53
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c6467725-3cbb-4da4-9374-f87daf36fb9b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nmussy commented on packages/aws-cdk-lib/aws-codebuild/lib/fleet.ts:
```diff
@@ -138,13 +205,38 @@ export class Fleet extends Resource implements IFleet {
+
+    this.validatePositiveInteger(diskGiB, 'disk size');
+    this.validatePositiveInteger(memoryGiB, 'memory size');
+    if (!Token.isUnresolved(vCpu)) {
+      this.validatePositiveInteger(vCpu, 'vCPU count');
+    }
```
You could move the `Token.isUnresolved` check to `validatePositiveInteger`, it's a little cleaner and it can't hurt the other two to validate them as well

badmintoncryer replied:
You are absolutely right. I have made the corrections!

nmussy commented on packages/aws-cdk-lib/aws-codebuild/lib/fleet.ts:
```diff
@@ -125,6 +191,7 @@ export class Fleet extends Resource implements IFleet {
   public readonly environmentType: EnvironmentType;
 
   constructor(scope: Construct, id: string, props: FleetProps) {
+    super(scope, id, { physicalName: props.fleetName });
```
I just noticed this addition, thanks I'd missed the `physicalName` in my original implementation. Wouldn't something like this be preferable though, I don't see a reason not to give `Resource` the rest of the props?
```ts
super(scope, id, { ...props, physicalName: props.fleetName });
```

nmussy commented on packages/aws-cdk-lib/aws-codebuild/lib/fleet.ts:
```diff
@@ -138,13 +205,38 @@ export class Fleet extends Resource implements IFleet {
+    }
+
+    const computeConfig = props.computeConfiguration;
+    const diskGiB = computeConfig?.disk?.toGibibytes() ?? 0;
+    const memoryGiB = computeConfig?.memory?.toGibibytes() ?? 0;
+    const vCpu = computeConfig?.vCpu ?? 0;
```
Probably a good idea to add a comment to explain why we need to have these default values here as well

badmintoncryer replied:
Thanks. Your PR had an excellent comment, so I used it as it was.

```ts
    // Despite what the CloudFormation schema says, the numeric properties are not optional.
    // An undefined value will cause the CloudFormation deployment to fail, e.g.
    // > Cannot invoke "java.lang.Integer.intValue()" because the return value of "software.amazon.codebuild.fleet.ComputeConfiguration.getMemory()" is null
```

nmussy commented on packages/aws-cdk-lib/aws-codebuild/lib/fleet.ts:
```diff
@@ -35,6 +35,69 @@ export interface FleetProps {
+  /**
+   * The amount of disk space of the instance type included in your fleet.
+   *
+   * @default 0
```
I don't know if it's necessary to tell the user we set the default value to 0, same with the CloudFormation bug explanation in the JSDoc. It's a bit confusing and shouldn't affect them. Despite the workaround, in effect we still have an unset value and the instance is selected according to the other criteria

badmintoncryer replied:
Thanks. Your PR had an excellent comment, so I used it as it was.

```ts
    // Despite what the CloudFormation schema says, the numeric properties are not optional.
    // An undefined value will cause the CloudFormation deployment to fail, e.g.
    // > Cannot invoke "java.lang.Integer.intValue()" because the return value of "software.amazon.codebuild.fleet.ComputeConfiguration.getMemory()" is null
```

badmintoncryer replied:
You're right. I also borrowed the description from your PR.

nmussy reviewed (changes requested):
More nits, sorry 😅

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 629680b1ffa171d46765b315d73d3b66956aebd1
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6ff03b72-4a80-4bea-b78e-80eb9e453719%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 6c45f77c8faa5af14f4e23c70637220a0cf6b461
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b01a70a3-4f94-4e8a-b0f9-888927b179e8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nmussy reviewed (approved):
LGTM, thanks for the changes 👍

badmintoncryer commented:
@nmussy Thanks a lot!!!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: cec7cfa965557c337b73feb21fa6b8f7b5be8dad
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ed38c186-3196-4d47-887a-66d1046ab3cf%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 6a1e46f80bc5f82da3ab870c16b718b6d3645414
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=179afead-e9cc-4c12-b7d3-3cac3ddb8888%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

GavinZZ commented on packages/@aws-cdk-testing/framework-integ/test/aws-codebuild/test/integ.project-fleet-attribute-based-compute.ts:
```diff
@@ -0,0 +1,77 @@
+  testCases: stacks,
+});
+
+const listFleets = test.assertions.awsApiCall('Codebuild', 'listFleets');
+for (const { fleet, project } of stacks) {
+  listFleets.expect(integ.ExpectedResult.objectLike({
+    fleets: integ.Match.arrayWith([fleet.fleetArn]),
+  }));
+
+  const startBuild = test.assertions.awsApiCall('Codebuild', 'startBuild', { projectName: project.projectName });
```
Nice assertions!

GavinZZ reviewed (approved):
LGTM, thank you as always!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 1b4ef4f318e9be8ee1b601ad04f18fe7de0f3ecd
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e81dae89-0c55-46c1-9dab-603bad4e9108%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: a65cafa7cbb90f088dc6a6b211049e8df10e99ca
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=65832688-3815-471d-8820-f5a41541a344%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 3c26b93a4ad34b539d74bfc14618303c186cd6c9
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=58774bde-2093-442d-942e-38100f41657e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 91cba4475b8ce4934a7dd903af59ac1bf3ceea72
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=084defcb-96b2-4264-9800-bd8bb7910079%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

mrgrain commented:
@mergifyio dequeue

mrgrain commented:
@mergifyio requeue

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 350b15911fc2a401db6ad375d20e86ee8a0e15c2
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5854f3df-02cf-45a6-a8ea-5871a1280b69%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
